packages/flutter_tools/lib/src/mdns_discovery.dart

```diff
@@ -0,0 +1,172 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:multicast_dns/multicast_dns.dart';
+
+import 'base/common.dart';
+import 'base/context.dart';
+import 'base/io.dart';
+import 'device.dart';
+import 'globals.dart';
+
+/// A wrapper around [MDnsClient] to find a Dart observatory instance.
+class MDnsObservatoryDiscovery {
+  /// Creates a new [MDnsObservatoryDiscovery] object.
+  ///
+  /// The [client] parameter will be defaulted to a new [MDnsClient] if null.
+  /// The [applicationId] parameter may be null, and can be used to
+  /// automatically select which application to use if multiple are advertising
+  /// Dart observatory ports.
+  MDnsObservatoryDiscovery({MDnsClient mdnsClient})
+    : client = mdnsClient ?? MDnsClient();
+
+  /// The [MDnsClient] used to do a lookup.
+  final MDnsClient client;
+
+  static const String dartObservatoryName = '_dartobservatory._tcp.local';
```

Comment:
Does this need to be public? Could it be `@visibleForTesting`?

Reply:
added `@visibleForTesting`

Reply:
added `@visibleForTesting`